### PR TITLE
Deprecate conventions

### PR DESCRIPTION
**Pull request by donat**
The concept of conventions is outdated and superseded by extensions. This PR deprecates the following:
- `org.gradle.api.Project.getConvention()`
- `org.gradle.api.internal.HasConvention`

There several popular third-party plugins that make use of conventions. Because of that, we decided to only add the `@Deprecated` annotation to the items above, and to the related extensions methods in the Kotlin DSL.

We'll [introduce](https://github.com/gradle/gradle/issues/17176) deprecation logging messages in a later Gradle 7.x release.

### CODE REVIEW

**Comment by donat** [comment]
@bot-gradle test qF

**Comment by donat** [comment]
@bot-gradle test rFM

**Comment by donat** [comment]
@bot-gradle test rFM

**Comment by donat** [comment]
@bot-gradle test rFM

File: subprojects/core/src/main/java/org/gradle/api/internal/project/DefaultProject.java [hunk 1]
```diff
@@ -562,7 +562,15 @@ public void setConfigurationContainer(ConfigurationContainer configurationContai
...
-        // TODO (donat) deprecate after all internal usages have been eliminated
+        DeprecationLogger.deprecateMethod(Project.class, "getConvention()")
+            .replaceWith("Project.getExtensions()")
+            .willBeRemovedInGradle8()
+            .withUpgradeGuideSection(7, "project_getconvention_deprecation")
+            .nagUser();
+        return getInternalConvention();
+    }
+
+    public Convention getInternalConvention() {
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/project/DefaultProject.java:573]: I would probably just explicitly call `extensibleDynamicObject.getConvention()` in both places vs adding another method to remove later.

**Comment by donat** [comment]
@bot-gradle test rFM

**Comment by donat** [comment]
@bot-gradle test rFM

**Comment by donat** [comment]
@bot-gradle

**Comment by donat** [comment]
@bot-gradle test QuickFeedbackLinux

**Comment by donat** [comment]
@bot-gradle test QuickFeedbackLinux

**Comment by donat** [comment]
@bot-gradle test rFM

**Comment by donat** [comment]
@bot-gradle test QuickFeedbackLinux

**Comment by donat** [comment]
@bot-gradle test rFM

**Comment by donat** [comment]
@bot-gradle test rFM

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/ExecutionGlobalServices.java [hunk 2]
```diff
@@ -179,6 +179,7 @@ TypeAnnotationMetadataStore createAnnotationMetadataStore(CrossBuildInMemoryCach
             IGNORED_METHOD_ANNOTATIONS,
             method -> method.isAnnotationPresent(Generated.class),
             cacheFactory);
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/internal/service/scopes/ExecutionGlobalServices.java:181]: 💄 To make it more clear what is using a deprecated member I would extract some `deprecatedHasConventionClass` variable and annotate just that statement with `@SuppressWarnings` in the same way as `deprecatedAbstractTask` above.

File: subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/precompiled/PrecompiledScriptPlugins.kt [hunk 3]
```diff
@@ -78,4 +78,5 @@ val Project.sourceSets
 
 private
 val SourceSet.kotlin: SourceDirectorySet
-    get() = withConvention(KotlinSourceSet::class) { kotlin }
+    // TODO coordinate with the kotlin-gradle-plugin's authors to migrate from conventions to extensions
```
Comment (eskatos) [inline subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/precompiled/PrecompiledScriptPlugins.kt:81]: 💄 I would remove that comment and track this elsewhere

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/ProjectExtensions.kt [hunk 4]
```diff
@@ -81,10 +81,14 @@ inline fun <reified T : Plugin<Project>> Project.apply() =
...
+ * Note, that the concept of conventions is deprecated and scheduled for
+ * removal in Gradle 8.
+ *
  * @param T the plugin convention type.
  * @param configuration the configuration block.
  * @see [Convention.getPlugin]
  */
+@Suppress("deprecation")
 inline fun <reified T : Any> Project.configure(noinline configuration: T.() -> Unit): Unit =
     typeOf<T>().let { type ->
```
Comment (eskatos) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/ProjectExtensions.kt:93]: 💄 Move the suppression closer to the deprecated member usage
```suggestion
inline fun <reified T : Any> Project.configure(noinline configuration: T.() -> Unit): Unit =
    @Suppress("deprecation")
    typeOf<T>().let { type ->
```

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/ProjectExtensions.kt [hunk 5]
```diff
@@ -106,7 +114,11 @@ inline fun <reified T : Any> Project.the(): T =
 
 /**
  * Returns the plugin convention or extension of the specified type.
+ *
+ * Note, that the concept of conventions is deprecated and scheduled for
+ * removal in Gradle 8.
  */
+@Suppress("deprecation")
 fun <T : Any> Project.the(extensionType: KClass<T>): T =
     convention.findByType(extensionType.java)
```
Comment (eskatos) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/ProjectExtensions.kt:123]: 💄 Move the suppression closer to the deprecated member usage
```suggestion
fun <T : Any> Project.the(extensionType: KClass<T>): T =
    @Suppress("deprecation")
    convention.findByType(extensionType.java)
```

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/accessors/AccessorFragments.kt [hunk 6]
```diff
@@ -713,7 +713,7 @@ fun fragmentsForConvention(accessor: Accessor.ForConvention): Fragments {
     val (kotlinConventionType, jvmConventionType) = accessibleTypesFor(conventionType)
 
     return className to sequenceOf(
-
+        @Suppress("deprecation")
```
Comment (eskatos) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/accessors/AccessorFragments.kt:716]: :x: this suppression is not needed

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/accessors/runtime/Runtime.kt [hunk 7]
```diff
@@ -34,17 +33,21 @@ fun extensionOf(target: Any, extensionName: String): Any =
     (target as ExtensionAware).extensions.getByName(extensionName)
 
 
+@Deprecated("The concept of conventions is deprecated. Use extensions if possible.")
```
Comment (eskatos) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/accessors/runtime/Runtime.kt:36]: ❓ this type is internal and its members are invoked by generated bytecode only, I would not add `@Deprecated` to those three functions

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/accessors/CodeGenerator.kt [hunk 8]
```diff
@@ -109,12 +108,16 @@ fun accessibleConventionAccessorFor(targetType: String, name: AccessorNameSpec,
     """
         /**
          * Retrieves the [$original][$type] convention.
+         *
+         * @deprecated Use extensions instead.
```
Comment (eskatos) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/accessors/CodeGenerator.kt:112]: 💄 Let the deprecation documentation be more explicit and consistent with other places
```suggestion
         * @deprecated The concept of conventions is deprecated. Use extensions instead.
```
This applies to the other generated convention accessors below.

**Review by eskatos — CHANGES_REQUESTED** [review]
It's great to see conventions deprecated. I left some comments and suggestions, mostly minor things.

File: subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/precompiled/PrecompiledScriptPlugins.kt (hunk 3 above)
Comment (donat) [reply subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/precompiled/PrecompiledScriptPlugins.kt:81]: Ok, I'll remove the comment and follow-up elsewhere.

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/accessors/runtime/Runtime.kt (hunk 7 above)
Comment (donat) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/accessors/runtime/Runtime.kt:36]: Makes sense, thanks for the insight.

**Comment by donat** [comment]
@eskatos I've resolved all your review items. PTAL.

**Comment by donat** [comment]
@bot-gradle test rFM

**Review by eskatos — APPROVED** [review]
💯

**Pull request merged**